code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,201 @@ RSF
 	else
-		matter--
-		to_chat(user, "<span class='notice'>The RSF now holds [matter]/30 fabrication-units.</span>")
-
-/obj/item/cookiesynth
+		if(matter - 1 >= 0)
+			matter--
+			to_chat(user, "<span class='notice'>\The [src] now holds [matter]/[max_matter] [discriptor].</span>")
+			return TRUE
+		to_chat(user, "<span class='warning'>\The [src] doesn't have enough [discriptor] left.</span>")
+	icon_state = spent_icon_state
+	return FALSE
+
+///Helper proc that iterates through all the things we are allowed to spawn on, and sees if the passed atom is one of them
+/obj/item/rsf/proc/is_allowed(atom/to_check)
+	for(var/sort in allowed_surfaces)
+		if(istype(to_check, sort))
+			return TRUE
+	return FALSE
+
+/obj/item/rsf/cookiesynth
 	name = "Cookie Synthesizer"
 	desc = "A self-recharging device used to rapidly deploy cookies."
-	icon = 'icons/obj/tools.dmi'
 	icon_state = "rcd"
-	lefthand_file = 'icons/mob/inhands/equipment/tools_lefthand.dmi'
-	righthand_file = 'icons/mob/inhands/equipment/tools_righthand.dmi'
-	var/matter = 10
-	var/toxin = 0
+	spent_icon_state = "rcd"
+	var/toxin = FALSE
 	var/cooldown = 0
 	var/cooldowndelay = 10
-	w_class = WEIGHT_CLASS_NORMAL
+	max_matter = 10
+	cost_by_item = list()
+	dispense_cost = 100
+	discriptor = "cookie-units"
+	action_type = "Fabricates"
 
-/obj/item/cookiesynth/examine(mob/user)
+/obj/item/rsf/cookiesynth/Initialize()
 	. = ..()
-	. += "<span class='notice'>It currently holds [matter]/10 cookie-units.</span>"
+	START_PROCESSING(SSprocessing, src)
 
-/obj/item/cookiesynth/attackby()
+/obj/item/rsf/cookiesynth/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+/obj/item/rsf/cookiesynth/attackby()
 	return
 
-/obj/item/cookiesynth/emag_act(mob/user)
+/obj/item/rsf/cookiesynth/emag_act(mob/user)
 	obj_flags ^= EMAGGED
 	if(obj_flags & EMAGGED)
 		to_chat(user, "<span class='warning'>You short out [src]'s reagent safety checker!</span>")
 	else
 		to_chat(user, "<span class='warning'>You reset [src]'s reagent safety checker!</span>")
-		toxin = 0
 
-/obj/item/cookiesynth/attack_self(mob/user)
+/obj/item/rsf/cookiesynth/attack_self(mob/user)
 	var/mob/living/silicon/robot/P = null
 	if(iscyborg(user))
 		P = user
-	if((obj_flags & EMAGGED)&&!toxin)
-		toxin = 1
-		to_chat(user, "<span class='alert'>Cookie Synthesizer hacked.</span>")
-	else if(P.emagged&&!toxin)
-		toxin = 1
+	if(((obj_flags & EMAGGED) || (P && P.emagged)) && !toxin)
+		toxin = TRUE
+		to_dispense = /obj/item/reagent_containers/food/snacks/cookie/sleepy
 		to_chat(user, "<span class='alert'>Cookie Synthesizer hacked.</span>")
 	else
-		toxin = 0
+		toxin = FALSE
+		to_dispense = /obj/item/reagent_containers/food/snacks/cookie
 		to_chat(user, "<span class='notice'>Cookie Synthesizer reset.</span>")
 
-/obj/item/cookiesynth/process()
-	if(matter < 10)
-		matter++
+/obj/item/rsf/cookiesynth/process()
+	matter = min(matter + 1, max_matter) //We add 1 up to a point
```

Comment:
Could you make it stop processing once it reaches max capacity, and restart when it goes under max?

Reply:
Sure